lib/Scheduler/Scheduler.cpp

```diff
@@ -0,0 +1,97 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Scheduler/Scheduler.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "ttmlir/Dialect/TTIR/IR/TTIROpsDialect.h.inc"
+#include <mlir/IR/BuiltinOps.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/Interfaces/DestinationStyleOpInterface.h>
+
+namespace mlir::tt::scheduler {
+
+bool isTTIROp(mlir::Operation *op) {
+  return isa<ttir::TTIRDialect>(op->getDialect());
+}
+
+bool isDpsOp(mlir::Operation *op) {
+  return isa<mlir::DestinationStyleOpInterface>(op);
+}
+
+// Init the dependencies map of all TTIR ops
+// which implement DestinationStyleOpInterface
+// We check for DPS because all operations which do
+// some computation in TTIR implement this interface
+Scheduler::Scheduler(func::FuncOp *func) {
+  for (auto &op : func->getOps()) {
+    if (isTTIROp(&op) && isDpsOp(&op)) {
+      dependencies[&op] = {};
+      unscheduledOps.insert(&op);
+    }
+  }
+
+  for (auto &op : func->getOps()) {
+    // Skip non TTIR operations
+    // Skip operations which do not implement DestinationStyleOpInterface
+    if (!isTTIROp(&op) || !isDpsOp(&op)) {
+      continue;
+    }
+
+    auto dpsOp = cast<mlir::DestinationStyleOpInterface>(&op);
+    assert(dpsOp.getNumDpsInits() == 1 &&
+           "Operation must have a single DPS operand");
+    OpResult result = op.getResult(0);
+
+    for (mlir::Operation *use : result.getUsers()) {
+      // Skip non TTIR operations
+      // Skip operations which set the result
+      if (isTTIROp(use) && use->getResult(0) != result) {
+        dependencies[use].push_back(&op);
+      }
+    }
+  }
+}
+
+Scheduler::Scheduler(const Scheduler &scheduler)
+    : scheduledOpsMap(scheduler.scheduledOpsMap), schedule(scheduler.schedule),
+      unscheduledOps(scheduler.unscheduledOps),
+      dependencies(scheduler.dependencies) {}
+
+llvm::SmallVector<mlir::Operation *> Scheduler::getScheduleableOps() {
+  llvm::SmallVector<mlir::Operation *> scheduleableOps;
```

Comment:
Nit: You could also keep this as part of Scheduler state in order to avoid iterating over all unscheduled ops. Might be a premature optimization tho.

Reply:
I had that implementation for pybuda scheduler. I didn't like it it wasn't readable. If it proves that this is too slow we can change